File: tech-studio-projects/financial-industry/algo-trading-platform/cpp/logging/AbstractLogger.cpp
#include "AbstractLogger.h"

// Default implementation of logging methods in the base class.
void ILogger::Debug(const std::string& module, const std::string& message) {
    Log(LogLevel::Debug, module, message);
}

void ILogger::Debug(const std::string& module, const std::exception& exception) {
    Log(LogLevel::Debug, module, std::string{ exception.what() });
}

void ILogger::Information(const std::string& module, const std::string& message) {
    Log(LogLevel::Information, module, message);
}

void ILogger::Information(const std::string& module, const std::exception& exception) {
    Log(LogLevel::Information, module, std::string{ exception.what() });
}

void ILogger::Warning(const std::string& module, const std::string& message) {
    Log(LogLevel::Warning, module, message);
}

void ILogger::Warning(const std::string& module, const std::exception& exception) {
    Log(LogLevel::Warning, module, std::string{ exception.what() });
}

void ILogger::Error(const std::string& module, const std::string& message) {
    Log(LogLevel::Error, module, message);
}

void ILogger::Error(const std::string& module, const std::exception& exception) {
    Log(LogLevel::Error, module, std::string{ exception.what() });
}

File: tech-studio-projects/financial-industry/algo-trading-platform/java/algo-trading-platform/src/com/fintech/algotrading/logging/AbstractLogger.java
package com.fintech.algotrading.logging;

public abstract class AbstractLogger implements ILogger {

    public void debug(String module, String message) {
        log(LogLevel.DEBUG, module, message);
    }

    public void debug(String module, Exception exception) {
        log(LogLevel.DEBUG, module, exception.getMessage());
    }

    public void information(String module, String message) {
        log(LogLevel.INFORMATION, module, message);
    }

    public void information(String module, Exception exception) {
        log(LogLevel.INFORMATION, module, exception.getMessage());
    }

    public void warning(String module, String message) {
        log(LogLevel.WARNING, module, message);
    }

    public void warning(String module, Exception exception) {
        log(LogLevel.WARNING, module, exception.getMessage());
    }

    public void error(String module, String message) {
        log(LogLevel.ERROR, module, message);
    }

    public void error(String module, Exception exception) {
        log(LogLevel.ERROR, module, exception.getMessage());
    }

    // Abstract method to be implemented by concrete classes
    protected abstract void log(LogLevel level, String module, String message);
}

